src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.StateSet.cs

```diff
@@ -103,29 +120,24 @@ public IEnumerable<ProjectId> GetProjectsWithDiagnostics()
                                                            .Select(kv => kv.Key)));
             }
 
-            public IEnumerable<DocumentId> GetDocumentsWithDiagnostics(ProjectId projectId)
+            [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/34761", AllowCaptures = false, AllowGenericEnumeration = false)]
+            public void CollectDocumentsWithDiagnostics(ProjectId projectId, HashSet<DocumentId> set)
             {
-                HashSet<DocumentId> set = null;
-                foreach (var state in GetActiveFileStates(projectId))
+                Debug.Assert(set != null);
+
+                // Collect active documents with diagnostics
+                foreach (var kvp in _activeFileStates)
```

Comment:
There's still an allocation happening here, if we want to switch from ConcurrentDictionary to lock + Dictionary, we can get rid of it.